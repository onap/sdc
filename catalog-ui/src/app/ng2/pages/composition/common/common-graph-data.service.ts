import {Injectable} from "@angular/core";
import 'rxjs/add/observable/forkJoin';
import {ComponentInstance} from "../../../../models/componentsInstances/componentInstance";
import {SelectedComponentType} from "./store/graph.actions";
import {RelationshipModel} from "../../../../models/graph/relationship";

@Injectable()
export class CommonGraphDataService {

    public componentInstances: Array<ComponentInstance>;
    public componentInstancesRelations: RelationshipModel[];
    public selectedComponentType: SelectedComponentType;

    constructor() {
    }

    //------------------------ RELATIONS ---------------------------------//
    public setRelations = (componentInstancesRelations: RelationshipModel[]) => {
        this.componentInstancesRelations = this.componentInstancesRelations;
    }

    public getRelations = (): RelationshipModel[] => {
        return this.componentInstancesRelations;
    }

    public addRelation = (componentInstancesRelations: RelationshipModel) => {
        this.componentInstancesRelations.push(componentInstancesRelations);
    }

    public deleteRelation(relationToDelete: RelationshipModel) {
        this.componentInstancesRelations = _.filter(this.componentInstancesRelations, (relationship: RelationshipModel) => {
            return relationship.relationships[0].relation.id !== relationToDelete.relationships[0].relation.id;
        });
    }

    //---------------------------- COMPONENT INSTANCES ------------------------------------//
    public getComponentInstances = (): Array<ComponentInstance> => {
        return this.componentInstances;
    }

    public addComponentInstance = (instance: ComponentInstance) => {
        return this.componentInstances.push(instance);
    }

    public updateComponentInstances = (componentInstances: ComponentInstance[]) => {
        _.unionBy(this.componentInstances, componentInstances, 'uniqueId');
    }

    public updateInstance = (instance: ComponentInstance) => {
        this.componentInstances = this.componentInstances.map(componentInstance => instance.uniqueId === componentInstance.uniqueId? instance : componentInstance);
    }

    public deleteComponentInstance(instanceToDelete: string) {
        this.componentInstances = _.filter(this.componentInstances, (instance: ComponentInstance) => {
            return instance.uniqueId !== instanceToDelete;
        });
    }

    //----------------------------SELECTED COMPONENT -----------------------//

    public setSelectedComponentType = (selectedType: SelectedComponentType) => {
        this.selectedComponentType = selectedType;
    }
}
