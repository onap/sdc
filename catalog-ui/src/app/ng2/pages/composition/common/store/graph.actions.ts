export enum SelectedComponentType {
    COMPONENT_INSTANCE = "COMPONENT_INSTANCE",
    GROUP = "GROUP",
    POLICY = "POLICY",
    TOPOLOGY_TEMPLATE = "TOPOLOGY_TEMPLATE"
}

export class UpdateSelectedComponentAction {
    static readonly type = '[COMPOSITION] UpdateSelectedComponent';

        constructor(public payload: {uniqueId?: string, type?: string}) {
    }
}
    
export class SetSelectedComponentAction {
    static readonly type = '[COMPOSITION] SetSelectedComponent';

    constructor(public payload: {component?: any, type?: SelectedComponentType}) {
        }
}

export class OnSidebarOpenOrCloseAction {
    static readonly type = '[COMPOSITION] OnSidebarOpenOrCloseAction';

    constructor() {
    }
}

export class TogglePanelLoadingAction {
    static readonly type = '[COMPOSITION] TogglePanelLoading';
    constructor(public payload: { isLoading: boolean}) {
    }
}
