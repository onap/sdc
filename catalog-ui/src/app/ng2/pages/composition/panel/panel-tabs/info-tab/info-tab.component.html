<!--
  ~ Copyright (C) 2018 AT&T Intellectual Property. All rights reserved.
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~      http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  -->
 
<ng2-expand-collapse state="0">
    <header tooltip="General Information">General Info</header>
    <content class="general-info-container">
        <!-- TYPE -->
        <div class="component-details-panel-item" *ngIf="component.componentType">
            <span class="name" [innerHTML]="'Type:'"></span>
            <span class="value" data-tests-id="rightTab_componentType" tooltip="{{component.componentType}}">{{component.componentType}}</span>
        </div>

        <!-- RESOURCE TYPE-->
        <div class="component-details-panel-item" *ngIf="component.resourceType">
            <span class="name" [innerHTML]="'Resource Type:'"></span>
            <span class="value" data-tests-id="rightTab_resourceType" tooltip="{{component.resourceType}}">{{component.resourceType}}</span>
        </div>

        <!-- VERSION -->
        <div class="component-details-panel-item" >
            <span class="name" [innerHTML]="'GENERAL_LABEL_VERSION' | translate"></span>
            <span class="value" *ngIf="!isComponentSelectedFlag" data-tests-id="rightTab_version" tooltip="{{component.version}}">{{component.version}}</span>
            <ng-container *ngIf="isComponentSelectedFlag">
                <select #versionDropdown (change)="onChangeVersion(versionDropdown)" [ngModel]="component.getComponentUid()" data-tests-id="changeVersion">
                    <option *ngFor="let version of versions" value="{{version.value}}"
                            [disabled]="isDisabledFlag" [class.minor]="(component.componentVersion)%1"
                    >{{version.label}}</option>
                </select>
            </ng-container>
        </div>

        <!-- CATEGORY -->
        <ng-container *ngIf="component.categories && component.categories[0]">
            <div class="component-details-panel-item">
                <span class="name" [innerHTML]="'GENERAL_LABEL_CATEGORY' | translate"></span>
                <span class="value" data-tests-id="rightTab_category" tooltip="{{component.categories[0].name}}">{{component.categories[0].name}}</span>
            </div>

            <!-- SUB CATEGORY -->
            <div class="component-details-panel-item" *ngIf="component.categories[0].subcategories && component.categories[0].subcategories[0]">
                <span class="name" [innerHTML]="'GENERAL_LABEL_SUB_CATEGORY' | translate"></span>
                <span class="value" data-tests-id="rightTab_subCategory" tooltip="{{component.categories[0].subcategories[0].name}}">{{component.categories[0].subcategories[0].name}}</span>
            </div>
        </ng-container>

        <!-- CREATION DATE -->
        <div class="component-details-panel-item" *ngIf="component.creationDate">
            <span class="name" [innerHTML]="'Creation Date:'"></span>
            <span class="value" data-tests-id="rightTab_version" tooltip="{{component.creationDate | date: 'MM/dd/yyyy'}}">{{component.creationDate | date: 'MM/dd/yyyy'}}</span>
        </div>

        <!-- AUTHOR -->
        <div class="component-details-panel-item" *ngIf="component.creatorFullName">
            <span class="name" [innerHTML]="'Author:'"></span>
            <span class="value" data-tests-id="rightTab_author" tooltip="{{component.creatorFullName}}">{{component.creatorFullName}}</span>
        </div>

        <!-- Vendor Name data-ng-if="selectedComponent.isResource()"-->
        <div class="component-details-panel-item" *ngIf="component.vendorName">
            <span class="name" [innerHTML]="'Vendor Name:'"></span>
            <span class="value" data-tests-id="rightTab_vendorName" tooltip="{{component.vendorName}}">{{component.vendorName}}</span>
        </div>

        <!-- Vendor Release data-ng-if="selectedComponent.isResource()"-->
        <div class="component-details-panel-item" *ngIf="component.vendorRelease">
            <span class="name" [innerHTML]="'Vendor Release:'"></span>
            <span class="value" data-tests-id="rightTab_vendorRelease" tooltip="{{component.vendorRelease}}">{{component.vendorRelease}}</span>
        </div>

        <!-- Vendor Release data-ng-if="selectedComponent.isResource()"-->
        <div class="component-details-panel-item" *ngIf="component.resourceVendorModelNumber">
            <span class="name" [innerHTML]="'GENERAL_LABEL_RESOURCE_MODEL_NUMBER' | translate"></span>
            <span class="value" data-tests-id="rightTab_resourceVendorModelNumber" tooltip="{{component.resourceVendorModelNumber}}">{{component.resourceVendorModelNumber}}</span>
        </div>

        <!-- Service Type data-ng-if="selectedComponent.isService()"-->
        <div class="component-details-panel-item" *ngIf="component.serviceType">
            <span class="name" [innerHTML]="'GENERAL_LABEL_SERVICE_TYPE' | translate"></span>
            <span class="value" data-tests-id="rightTab_serviceType" tooltip="{{component.serviceType}}">{{component.serviceType}}</span>
        </div>

        <!-- Service Role data-ng-if="selectedComponent.isService()"-->
        <div class="component-details-panel-item" *ngIf="component.serviceRole">
            <span class="name" [innerHTML]="'GENERAL_LABEL_SERVICE_ROLE' | translate"></span>
            <span class="value" data-tests-id="rightTab_serviceRole" tooltip="{{component.serviceRole}}">{{component.serviceRole}}</span>
        </div>

        <!-- Contact ID -->
        <div class="component-details-panel-item" *ngIf="component.contactId">
            <span class="name" [innerHTML]="'GENERAL_LABEL_CONTACT_ID' | translate"></span>
            <span class="value" data-tests-id="rightTab_contactId" tooltip="{{component.contactId}}">{{component.contactId}}</span>
        </div>

        <!-- Service Name data-ng-if="isComponentInstanceSelected() && currentComponent.selectedInstance.isServiceProxy()"-->
        <div class="component-details-panel-item" *ngIf="component.sourceModelName">
            <span class="name" [innerHTML]="'GENERAL_LABEL_SOURCE_SERVICE_NAME' | translate"></span>
            <span class="value" data-tests-id="rightTab_sourceModelName" tooltip="{{component.sourceModelName}}">{{component.sourceModelName}}</span>
        </div>

        <!-- Customization UUID data-ng-if="isViewMode() && currentComponent.isService() && selectedComponent.isResource()"-->
        <div class="component-details-panel-item" *ngIf="component.customizationUUID">
            <span class="name" [innerHTML]="'GENERAL_LABEL_RESOURCE_CUSTOMIZATION_UUID' | translate"></span>
            <span class="value" data-tests-id="rightTab_customizationModuleUUID" tooltip="{{component.customizationUUID}}">{{component.customizationUUID}}</span>
        </div>

        <!-- DESCRIPTION -->
        <div class="component-details-panel-item description">
            <span class="name" [innerHTML]="'GENERAL_LABEL_DESCRIPTION' | translate"></span>
            <chars-ellipsis [text]="component.description" [maxChars]="55" [testId]="'rightTab_description'"></chars-ellipsis>
        </div>


        <!--TODO: move to separate component!-->
        <ng-container *ngIf="componentType == 'POLICY'">
            <!-- TYPE -->
            <div class="component-details-panel-item policy-item">
                <span class="name" [innerHTML]="'GENERAL_LABEL_TYPE' | translate"></span>
                <span class="value" data-tests-id="rightTab_componentType" tooltip="{{component.policyTypeUid}}">{{component.policyTypeUid}}</span>
            </div>

            <!-- CATEGORY -->
            <div class="component-details-panel-item policy-item">
                <span class="name" [innerHTML]="'GENERAL_LABEL_CATEGORY' | translate"></span>
                <span class="value" data-tests-id="rightTab_category" tooltip="Policy">Policy</span>
            </div>

            <!-- SUB CATEGORY -->
            <div class="component-details-panel-item policy-item">
                <span class="name" [innerHTML]="'GENERAL_LABEL_SUB_CATEGORY' | translate"></span>
                <span class="value" data-tests-id="rightTab_subCategory" tooltip="Policy">Policy</span>
            </div>
        </ng-container>

        <!--TODO: move to separate component!-->
        <ng-container *ngIf="componentType == 'GROUP'">
            <!-- CATEGORY -->
            <div class="component-details-panel-item group-item">
                <span class="name" [innerHTML]="'GENERAL_LABEL_CATEGORY' | translate"></span>
                <span class="value" data-tests-id="rightTab_category" tooltip="Group">Group</span>
            </div>

            <!-- SUB CATEGORY -->
            <div class="component-details-panel-item group-item">
                <span class="name" [innerHTML]="'GENERAL_LABEL_SUB_CATEGORY' | translate"></span>
                <span class="value" data-tests-id="rightTab_subCategory" tooltip="Group">Group</span>
            </div>

        </ng-container>

    </content>
</ng2-expand-collapse>

<ng2-expand-collapse *ngIf="component.tags || isComponentInstanceSelected()">
    <header tooltip="Tags">Tags</header>
    <content class="tags-container">
        <span *ngIf="component.tags?.indexOf(component.name)===-1" class="i-sdc-designer-sidebar-section-content-item-tag"
              data-tests-id="rightTab_tag" tooltip="{{component.name}}">{{component.name}}</span>
        <span class="i-sdc-designer-sidebar-section-content-item-tag" *ngFor="let tag of component.tags"
              data-tests-id="rightTab_tag" tooltip="{{tag}}">{{tag}}</span>
    </content>
</ng2-expand-collapse>
