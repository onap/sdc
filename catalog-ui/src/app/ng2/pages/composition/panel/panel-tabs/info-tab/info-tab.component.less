@import '../../../../../../../assets/styles/variables';

.general-info-container {
    display: flex;
    flex-direction: column;
    padding: 10px 20px;
}

.component-details-panel-item {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    margin-bottom: 5px;
    order:1;

    .name { font-family: OpenSans-Semibold, sans-serif; }
    .value { padding-left: 10px; }


    &.description {
        margin-top: 28px;
        white-space: normal;
        word-wrap: break-word;
        overflow: ellipsis;

        .value {
            padding-left: 0;
            max-width: none;
            font-weight: normal;
            font-family: @font-opensans-regular;
        }
    }

    &.group-item, &.policy-item {
        order:0;
    }
}

.tags-container {
    display: flex;
    flex-wrap: wrap;
    padding: 10px 20px;

    .i-sdc-designer-sidebar-section-content-item-tag {
        padding: 5px;
        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
        user-select: all;
    }
}
