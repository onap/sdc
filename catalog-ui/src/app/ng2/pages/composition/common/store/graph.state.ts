import { Action, Selector, State, StateContext} from '@ngxs/store';
import {
    OnSidebarOpenOrCloseAction,
    SelectedComponentType,
    SetSelectedComponentAction,
    TogglePanelLoadingAction
} from "./graph.actions";
import { PolicyInstance, GroupInstance, Component as TopologyTemplate, ComponentInstance, LeftPaletteComponent, FullComponentInstance} from "app/models";
import { TopologyTemplateService } from "app/ng2/services/component-services/topology-template.service";
import { tap } from "rxjs/operators";
import { CompositionService } from "app/ng2/pages/composition/composition.service";
import {GroupsService} from "../../../../services/groups.service";
import {PoliciesService} from "../../../../services/policies.service";
import {WorkspaceService} from "../../../workspace/workspace.service";

export class CompositionStateModel {

    isViewOnly?: boolean;
    panelLoading?: boolean;
    selectedComponentType?: SelectedComponentType;
    selectedComponent?: PolicyInstance | GroupInstance | TopologyTemplate | ComponentInstance;
    withSidebar?: boolean;
}

@State<CompositionStateModel>({
    name: 'composition',
    defaults: {
         withSidebar: true
    }
})
export class GraphState {

    constructor(private topologyTemplateService: TopologyTemplateService,
                private compositionService: CompositionService,
         private policiesService:PoliciesService, private groupsService:GroupsService,
                 private workspaceService: WorkspaceService) {}

    @Action(SetSelectedComponentAction)
    setSelectedComponent({dispatch, getState, patchState}:StateContext<CompositionStateModel>, action: SetSelectedComponentAction) {
        
        const state:CompositionStateModel = getState();
        
        patchState({ panelLoading: true });

        if(action.payload.component instanceof ComponentInstance){
            let originComponent = this.compositionService.getOriginComponentById(action.payload.component.getComponentUid());
            if(!originComponent) {
                return this.topologyTemplateService.getFullComponent(action.payload.component.originType, action.payload.component.getComponentUid())
                .pipe(tap(resp => {
                    this.compositionService.addOriginComponent(resp);
                    this.compositionService.setSelectedComponentType(SelectedComponentType.COMPONENT_INSTANCE);
                    patchState({
                        selectedComponent: new FullComponentInstance(action.payload.component, resp), 
                        selectedComponentType: action.payload.type,
                        panelLoading: false
                    });
                    }, err => {
                        patchState({
                            panelLoading: false
                        })
                    }
                ));
            } else {
                patchState({
                    selectedComponent: new FullComponentInstance(action.payload.component, originComponent), 
                    selectedComponentType: action.payload.type,
                    panelLoading: false
                });
            }
        } else if (action.payload.component instanceof PolicyInstance) {
            let topologyTemplate = this.workspaceService.metadata;
            return this.policiesService.getSpecificPolicy(topologyTemplate.componentType, topologyTemplate.uniqueId, action.payload.component.uniqueId).pipe(tap(resp => 
                {
                    this.compositionService.updatePolicy(resp);
                    patchState({
                        selectedComponent: resp, 
                        selectedComponentType: action.payload.type,
                        panelLoading: false
                    })
                }, err => {
                    patchState({
                        panelLoading: false
                    })
                 }
            ));
        
        } else if (action.payload.component instanceof GroupInstance) {
            let topologyTemplate = this.workspaceService.metadata;
            return this.groupsService.getSpecificGroup(topologyTemplate.componentType, topologyTemplate.uniqueId, action.payload.component.uniqueId).pipe(tap(resp => {
                this.compositionService.updateGroup(resp);
                patchState({
                    selectedComponent: resp,
                    selectedComponentType: action.payload.type,
                    panelLoading: false
                });
            }, err => {
                patchState({
                    panelLoading: false
                })
             }
            ));
        } else { //TopologyTemplate
            patchState({
                selectedComponent: action.payload.component,
                selectedComponentType: action.payload.type,
                panelLoading: false
            })
        }
    }


    // @Action(UpdateSelectedComponentNameAction)
    // UpdateSelectedComponentNameAction({patchState}:StateContext<CompositionStateModel>, action: UpdateSelectedComponentNameAction) {

    //     switch(action.payload.type){
    //         case SelectedComponentType.COMPONENT_INSTANCE:
    //             this.store.dispatch(new UpdateComponentInstancesAction([action.payload.component]));
    //             break;
    //         case SelectedComponentType.POLICY:
    //             this.store.dispatch(new UpdatePolicyNameAction(action.payload.uniqueId, action.payload.newName));
    //             break;
    //         case SelectedComponentType.GROUP:
    //             this.store.dispatch(new UpdateGroupInstancesAction)

    //     }
    //     if(action.payload.type === SelectedComponentType.COMPONENT_INSTANCE){
            
    //     } 
    
    // }

    @Selector() 
    static getSelectedComponent(state:CompositionStateModel) {
        return state.selectedComponent;
    }

    @Selector() 
    static getSelectedComponentId(state:CompositionStateModel) {
        return state.selectedComponent.uniqueId;
    }

    @Selector() 
    static getSelectedComponentType(state:CompositionStateModel) {
        return state.selectedComponentType;
    }


    @Action(OnSidebarOpenOrCloseAction)
    onSidebarOpenOrCloseAction({getState, setState}:StateContext<CompositionStateModel>) {
        const state:CompositionStateModel = getState();

        setState({
            ...state,
            withSidebar: !state.withSidebar
        });
    }
    
    @Action(TogglePanelLoadingAction)
    TogglePanelLoading({patchState}:StateContext<CompositionStateModel>, action: TogglePanelLoadingAction) {

        patchState({
            panelLoading: action.payload.isLoading
        });
    }

    @Selector() static withSidebar(state):boolean {
        return state.withSidebar;
    }

}