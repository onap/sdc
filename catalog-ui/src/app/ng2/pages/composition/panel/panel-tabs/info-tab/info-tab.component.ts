import { Component, OnInit, Input, Inject, OnDestroy } from '@angular/core';
import {
    PolicyInstance,
    GroupInstance,
    Component as TopologyTemplate,
    ComponentInstance,
    LeftPaletteComponent,
    FullComponentInstance
} from "app/models";
import {Store} from "@ngxs/store";
import { EVENTS, GRAPH_EVENTS } from 'app/utils';
import {IDropDownOption} from "onap-ui-angular/dist/form-elements/dropdown/dropdown-models";
import { CompositionPaletteService } from "app/ng2/pages/composition/palette/services/palette.service";
import { SdcUiCommon, SdcUiComponents, SdcUiServices } from "onap-ui-angular";
import { SdcMenuToken, IAppMenu } from "app/ng2/config/sdc-menu.config";
import { CompositionService } from "app/ng2/pages/composition/composition.service";
import { ServiceServiceNg2 } from "app/services-ng2";
import { WorkspaceService } from "app/ng2/pages/workspace/workspace.service";
import { ComponentInstanceServiceNg2 } from "app/ng2/services/component-instance-services/component-instance.service";
import { EventListenerService } from "app/services";
import * as _ from 'lodash';
import {SelectedComponentType, TogglePanelLoadingAction} from "../../../common/store/graph.actions";
import Dictionary = _.Dictionary;


@Component({
    selector: 'panel-info-tab',
    templateUrl: './info-tab.component.html',
    styleUrls: ['./info-tab.component.less'],
    // providers: [SdcUiServices.ModalService]
})
export class InfoTabComponent implements OnInit, OnDestroy {

    @Input() isViewOnly: boolean;
    @Input() componentType: SelectedComponentType;
    @Input() component: TopologyTemplate | PolicyInstance | GroupInstance | ComponentInstance;
    public versions: IDropDownOption[];
    private leftPalletElements: LeftPaletteComponent[];
    private isDisabledFlag: boolean;
    private isComponentSelectedFlag: boolean;

    constructor(private store: Store,
                private compositionPaletteService: CompositionPaletteService,
                private compositionService: CompositionService,
                private workspaceService: WorkspaceService,
                private modalService: SdcUiServices.ModalService,
                private componentInstanceService: ComponentInstanceServiceNg2,
                private serviceService: ServiceServiceNg2,
                private eventListenerService: EventListenerService,
                @Inject(SdcMenuToken) public sdcMenu:IAppMenu) {
    }

    ngOnInit() {
        this.leftPalletElements = this.flatLeftPaletteElementsFromService(this.compositionPaletteService.getLeftPaletteElements());
        this.initEditResourceVersion(this.component, this.leftPalletElements);
        this.eventListenerService.registerObserverCallback(EVENTS.ON_CHECKOUT, (comp) => {
            this.component = comp;
        });
        this.isComponentSelectedFlag = this.isComponentInstanceSelected();
        this.isDisabledFlag = this.isDisabled();

    }

    ngOnDestroy() {
        this.eventListenerService.unRegisterObserver(EVENTS.ON_CHECKOUT);
    }

    flatLeftPaletteElementsFromService = (leftPalleteElementsFromService: Dictionary<Dictionary<LeftPaletteComponent[]>>): LeftPaletteComponent[] => {
        let retValArr = [];
        for (const category in leftPalleteElementsFromService) {
            for (const subCategory in leftPalleteElementsFromService[category]) {
                retValArr = retValArr.concat(leftPalleteElementsFromService[category][subCategory].slice(0));
            }
        }
        return retValArr;
    }

    private isComponentInstanceSelected () {
        return this.componentType === SelectedComponentType.COMPONENT_INSTANCE;
    }

    private versioning: Function = (versionNumber: string): string => {
        let version: Array<string> = versionNumber && versionNumber.split('.');
        return '00000000'.slice(version[0].length) + version[0] + '.' + '00000000'.slice(version[1].length) + version[1];
    };


    private onChangeVersion = (versionDropdown) => {
        let newVersionValue = versionDropdown.value;
        versionDropdown.value = (<FullComponentInstance>this.component).getComponentUid();

        this.store.dispatch(new TogglePanelLoadingAction({isLoading: true}));

        // let service = <Service>this.$scope.currentComponent;
        if(this.component instanceof FullComponentInstance) {

            let onCancel = (error:any) => {
                this.store.dispatch(new TogglePanelLoadingAction({isLoading: false}));
                if (error) {
                    console.log(error);
                }
            };

            let onUpdate = () => {
                //this function will update the instance version than the function call getComponent to update the current component and return the new instance version
                this.componentInstanceService.changeResourceInstanceVersion(this.workspaceService.metadata.componentType, this.workspaceService.metadata.uniqueId, this.component.uniqueId, newVersionValue)
                    .subscribe((component) => {
                        this.store.dispatch(new TogglePanelLoadingAction({isLoading: false}));
                        this.eventListenerService.notifyObservers(GRAPH_EVENTS.ON_VERSION_CHANGED, component);
                    }, onCancel);
            };

            if (this.component.isService() || this.component.isServiceProxy() || this.component.isServiceSubstitution()) {
                this.serviceService.checkComponentInstanceVersionChange(this.workspaceService.metadata.componentType, this.workspaceService.metadata.uniqueId,
                    this.component.uniqueId, newVersionValue).subscribe((pathsToDelete:string[]) => {
                    if (pathsToDelete && pathsToDelete.length) {
                        this.store.dispatch(new TogglePanelLoadingAction({isLoading: false}));


                        const {title, message} = this.sdcMenu.alertMessages['upgradeInstance'];
                        let pathNames:string = this.getPathNamesVersionChangeModal(pathsToDelete);
                        let onOk: Function = () => {
                            this.store.dispatch(new TogglePanelLoadingAction({isLoading: true}));

                            onUpdate();
                        };
                        const okButton = {testId: "OK", text: "OK", type: SdcUiCommon.ButtonType.info, callback: onOk, closeModal: true} as SdcUiComponents.ModalButtonComponent;
                        const cancelButton = {testId: "Cancel", text: "Cancel", type: SdcUiCommon.ButtonType.secondary, callback: <Function>onCancel, closeModal: true} as SdcUiComponents.ModalButtonComponent;
                        const modal = this.modalService.openInfoModal(title, message.format([pathNames]), 'confirm-modal', [okButton, cancelButton]);
                        modal.getCloseButton().onClick(onCancel);
                    } else {
                        onUpdate();
                    }
                }, onCancel);
            } else {
                onUpdate();
            }
        }
    };


    private getPathNamesVersionChangeModal = (pathsToDelete:string[]):string => {
        const relatedPaths = _.filter(this.compositionService.forwardingPaths, path =>
            _.find(pathsToDelete, id =>
                path.uniqueId === id
            )
        ).map(path => path.name);
        const pathNames = _.join(relatedPaths, ', ') || 'none';
        return pathNames;
    };


    private initEditResourceVersion = (component, leftPaletteComponents): void => {
        if(this.component instanceof ComponentInstance) {

            this.versions = [];
            let sorted:any = _.sortBy(_.toPairs(component.allVersions), (item) => {
                return item[0] !== "undefined" && this.versioning(item[0]);
            });
            _.forEach(sorted, (item) => {
                this.versions.push({label: item[0], value: item[1]});
            });

            let highestVersion = _.last(sorted)[0];

            if (parseFloat(highestVersion) % 1) { //if highest is minor, make sure it is the latest checked in -
                let latestVersionComponent: LeftPaletteComponent = _.maxBy(
                    _.filter(leftPaletteComponents, (leftPaletteComponent: LeftPaletteComponent) => { //latest checked in
                        return (leftPaletteComponent.systemName === component.systemName || leftPaletteComponent.uuid === component.uuid);
                    })
                    , (component) => {
                        return component.version
                    });

                let latestVersion: string = latestVersionComponent ? latestVersionComponent.version : highestVersion;

                if (latestVersion && highestVersion != latestVersion) { //highest is checked out - remove from options
                    this.versions = this.versions.filter(version => version.label != highestVersion);
                }
            }
        }
    }

    private isDisabled() {
        return this.isViewOnly || this.component['archived'] || this.component['resourceType'] === 'CVFC'
    }

};

