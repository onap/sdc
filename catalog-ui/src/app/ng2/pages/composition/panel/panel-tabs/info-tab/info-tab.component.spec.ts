import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { Store } from '@ngxs/store';
import { CompositionPaletteService } from '../../../../../pages/composition/palette/services/palette.service';
import { IAppMenu, SdcMenuToken } from '../../../../../../../app/ng2/config/sdc-menu.config';
import { CompositionService } from '../../../../../pages/composition/composition.service';
import { ServiceServiceNg2 } from '../../../../../../../app/services-ng2';
import { WorkspaceService } from '../../../../../../../app/ng2/pages/workspace/workspace.service';
import { ComponentInstanceServiceNg2 } from '../../../../../../../app/ng2/services/component-instance-services/component-instance.service';
import { EventListenerService } from '../../../../../../../app/services';
import { InfoTabComponent } from './info-tab.component';
import { ConfigureFn, configureTests } from "../../../../../../../jest/test-config.helper";
import { Observable } from "rxjs";
import { leftPaletteElements } from "../../../../../../../jest/mocks/left-paeltte-elements.mock";
import { TranslatePipe } from "../../../../../shared/translator/translate.pipe";
import { HttpClientModule } from "@angular/common/http";
import { TranslateModule } from "../../../../../../../app/ng2/shared/translator/translate.module";
import _ from "lodash";
import { TranslateService } from "../../../../../shared/translator/translate.service";
import { SdcUiServices } from "onap-ui-angular";
import { Component as TopologyTemplate, FullComponentInstance, ComponentInstance } from '../../../../../../../app/models';


describe('InfoTabComponent', () => {
    // let comp: InfoTabComponent;
    let fixture: ComponentFixture<InfoTabComponent>;

    // let eventServiceMock: Partial<EventListenerService>;
    let storeStub:Partial<Store>;
    let compositionPaletteServiceStub:Partial<CompositionPaletteService>;
    let iAppMenuStub:Partial<IAppMenu>;
    let compositionServiceStub:Partial<CompositionService>;
    let serviceServiceNg2Stub:Partial<ServiceServiceNg2>;
    let workspaceServiceStub:Partial<WorkspaceService>;
    let componentInstanceServiceNg2Stub:Partial<ComponentInstanceServiceNg2>;
    let eventListenerServiceStub:Partial<EventListenerService>;

    beforeEach(
        async(() => {
            storeStub = {};
            iAppMenuStub = {};
            eventListenerServiceStub = {
                notifyObservers: jest.fn()
            }
            compositionPaletteServiceStub = {
                getLeftPaletteElements:  jest.fn().mockImplementation(()=>  Observable.of(leftPaletteElements))
            }
            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    imports: [ ],
                    declarations: [ InfoTabComponent, TranslatePipe ],
                    schemas: [ NO_ERRORS_SCHEMA ],
                    providers: [
                        { provide: Store, useValue: {} },
                        { provide: CompositionPaletteService, useValue: compositionPaletteServiceStub },
                        { provide: SdcMenuToken, useValue: {} },
                        { provide: CompositionService, useValue: {} },
                        { provide: SdcUiServices.ModalService, useValue: {}},
                        { provide: ServiceServiceNg2, useValue: {} },
                        { provide: WorkspaceService, useValue: {} },
                        { provide: ComponentInstanceServiceNg2, useValue: {} },
                        { provide: EventListenerService, useValue: eventListenerServiceStub },
                        { provide: TranslateService, useValue: {}}
                    ]
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(InfoTabComponent);
                let comp = fixture.componentInstance;

            });
        })
    );


    it('can load instance', () => {
        expect(fixture).toMatchSnapshot();
    });

    describe('Version dropdown', () => {
        it('is undefined for topologyTemplate', () => {
            fixture.componentInstance.component = <TopologyTemplate>{};
            fixture.componentInstance.initEditResourceVersion(fixture.componentInstance.component, fixture.componentInstance.flatLeftPaletteElementsFromService(leftPaletteElements));
            expect(fixture.componentInstance.versions).toBe(undefined);
        });
        it('does not contain the highest minor version if it is checked out', () => {
            fixture.componentInstance.component = new ComponentInstance();
            fixture.componentInstance.component.allVersions =  
            {'1.0': "9c829122-af05-4bc9-b537-5d84f4c8ae25", '1.1': "930d56cb-868d-4e35-bd0f-e737d2fdb171"};
            fixture.componentInstance.component.version = "1.0";
            fixture.componentInstance.component.uuid = "a8cf015e-e4e5-4d4b-a01e-8624e8d36095";
            fixture.componentInstance.initEditResourceVersion(fixture.componentInstance.component, fixture.componentInstance.flatLeftPaletteElementsFromService(leftPaletteElements));
            expect(fixture.componentInstance.versions).toHaveLength(1);
        });
    });

});
