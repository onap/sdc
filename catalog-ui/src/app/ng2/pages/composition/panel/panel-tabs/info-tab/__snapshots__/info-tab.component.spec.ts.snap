// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoTabComponent can load instance 1`] = `
<panel-info-tab
  componentInstanceService={[Function Object]}
  compositionPaletteService={[Function Object]}
  compositionService={[Function Object]}
  eventListenerService={[Function Object]}
  flatLeftPaletteElementsFromService={[Function Function]}
  getPathNamesVersionChangeModal={[Function Function]}
  initEditResourceVersion={[Function Function]}
  modalService={[Function Object]}
  onChangeVersion={[Function Function]}
  sdcMenu={[Function Object]}
  serviceService={[Function Object]}
  store={[Function Object]}
  versioning={[Function Function]}
  workspaceService={[Function Object]}
>
  <ng2-expand-collapse
    state="0"
  >
    <header
      tooltip="General Information"
    >
      General Info
    </header>
    <content
      class="general-info-container"
    >
      
      
      <div
        class="component-details-panel-item"
      >
        <span
          class="name"
        />
        
        
      </div>
      
      
      
      
      
      
      
      
      
      
      
      <div
        class="component-details-panel-item description"
      >
        <span
          class="name"
        />
        <chars-ellipsis />
      </div>
      
      
    </content>
  </ng2-expand-collapse>
</panel-info-tab>
`;
